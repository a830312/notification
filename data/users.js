import moment from 'moment'
var records = []


const getUsers = function(cb) {
  process.nextTick(function() {
    cb(records)
  })
}

const findById = function(id, cb) {
  process.nextTick(function() {
    var idx = id - 1;
    if (records[idx]) {
      cb(null, records[idx]);
    } else {
      cb(new Error('User ' + id + ' does not exist'))
    }
  })
}

const updateUser = function(username, cb) {
  process.nextTick(function() {
    findByUsername(username, function(err, user) {
      let numOfNotificationsPushed = user.numOfNotificationsPushed + 1,
          idx = user.id,
          newRecord = Object.assign({}, user, { numOfNotificationsPushed })
      records[idx] = newRecord
      cb(null, newRecord)
    })
  })
}

const findByUsername = function(username, cb) {
  process.nextTick(function() {
    let record = records.filter((record, i) => record.username === username)
    cb(null, record[0] || null)
  })
}

const addUser = function({username, accesstoken}, cb) {
  process.nextTick(function() {
    let user = {
      id: records.length,
      username,
      accessToken: accesstoken,
      creationTime: moment().format('YYYY-MM-DDTkk:mm:ss'),
      numOfNotificationsPushed: 0
    }
    records.push(user)
    cb(null, user)
  })
}

export default { findById, getUsers, updateUser, findByUsername, addUser }