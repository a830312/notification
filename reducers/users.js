import { actionTypes } from '../actions'

const initialState = []

const users = (state = initialState, action = {}) => {
  let { type, users } = action
  switch (type) {
    case actionTypes.LIST_USERS_SUCCESS:
      return users || state || initialState
    default:
      return state
  }
}

export default users