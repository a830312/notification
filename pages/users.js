import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import withRedux from 'next-redux-wrapper'
import { initStore } from '../reducers'
import { getUserList } from '../actions'
import User from '../components/user'
import Message from '../components/message'
import Head from '../components/head'
import Navigation from '../components/navigation'


class Users extends Component {

  static getInitialProps ({ store, isServer }) {
    store.dispatch(getUserList(isServer))

    return { isServer }
  }

  render() {
    let { users, message, title } = this.props
    return (
      <div>
        <Head title={`Notification - ${title}`} />
        <div className="container">
          <div className="col-sm-6 col-sm-offset-3">

              <h1>{ title }</h1>
              <Navigation />
              <Message message={ message } />

              {
                users.length ? (
                  (users || []).map((user, i) => 
                    <User user={user} key={`user-list-${i}`} />
                  )
                ) : <p>No users registered yet</p>
              }

          </div>
        </div>
      </div>
    )
  }
}

Users.propTypes = {
  users: PropTypes.array,
  message: PropTypes.string,
  title: PropTypes.string
}

Users.defaultProps = {
  title: 'List users'
}

const mapStateToProps = ({ users, message }) => {
  return {
    users,
    message
  }
}

const mapDispatchToProps = (dispatch) => {
  return {}
}

export { Users }
export default withRedux(initStore, mapStateToProps, mapDispatchToProps)(Users)

