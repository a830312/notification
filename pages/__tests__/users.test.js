import React from 'react'
import renderer from 'react-test-renderer'
import { Users } from '../users'


describe('Users Component', () => {


    it('should render with correct title', () => {
      const component = renderer.create(<Users title="List users" />)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render hint if no user registered yet', () => {
      const component = renderer.create(<Users title="List users" />)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should list users', () => {
      const users = [
        { username: 'foo', accessToken: '123' },
        { username: 'bar', accessToken: '456' }
      ]
      const component = renderer.create(<Users title="List users" users={users} />)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  
})

