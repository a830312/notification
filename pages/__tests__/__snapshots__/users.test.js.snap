// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Component should list users 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="col-sm-6 col-sm-offset-3 mx-auto"
    >
      <h1>
        List users
      </h1>
      <p>
        <a
          className="card-link"
          href="/"
          onClick={[Function]}
        >
          Register/Login
        </a>
        <a
          className="card-link"
          href="/logout"
          onClick={[Function]}
        >
          Logout
        </a>
        <a
          className="card-link"
          href="/pushes"
          onClick={[Function]}
        >
          Push Notification
        </a>
        <a
          className="card-link"
          href="/users"
          onClick={[Function]}
        >
          List users
        </a>
      </p>
      <div
        className="card form-group"
      >
        <div
          className="card-body"
        >
          <div
            className="card-text"
          >
            <div>
              <dt>
                username
              </dt>
              <dd>
                foo
              </dd>
            </div>
            <div>
              <dt>
                accessToken
              </dt>
              <dd>
                123
              </dd>
            </div>
          </div>
        </div>
      </div>
      <div
        className="card form-group"
      >
        <div
          className="card-body"
        >
          <div
            className="card-text"
          >
            <div>
              <dt>
                username
              </dt>
              <dd>
                bar
              </dd>
            </div>
            <div>
              <dt>
                accessToken
              </dt>
              <dd>
                456
              </dd>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Users Component should render hint if no user registered yet 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="col-sm-6 col-sm-offset-3 mx-auto"
    >
      <h1>
        List users
      </h1>
      <p>
        <a
          className="card-link"
          href="/"
          onClick={[Function]}
        >
          Register/Login
        </a>
        <a
          className="card-link"
          href="/logout"
          onClick={[Function]}
        >
          Logout
        </a>
        <a
          className="card-link"
          href="/pushes"
          onClick={[Function]}
        >
          Push Notification
        </a>
        <a
          className="card-link"
          href="/users"
          onClick={[Function]}
        >
          List users
        </a>
      </p>
      <p>
        No users registered yet
      </p>
    </div>
  </div>
</div>
`;

exports[`Users Component should render with correct title 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="col-sm-6 col-sm-offset-3 mx-auto"
    >
      <h1>
        List users
      </h1>
      <p>
        <a
          className="card-link"
          href="/"
          onClick={[Function]}
        >
          Register/Login
        </a>
        <a
          className="card-link"
          href="/logout"
          onClick={[Function]}
        >
          Logout
        </a>
        <a
          className="card-link"
          href="/pushes"
          onClick={[Function]}
        >
          Push Notification
        </a>
        <a
          className="card-link"
          href="/users"
          onClick={[Function]}
        >
          List users
        </a>
      </p>
      <p>
        No users registered yet
      </p>
    </div>
  </div>
</div>
`;
